report/node.go

```diff
@@ -74,121 +72,92 @@ func (n Node) After(other Node) bool {
 
 // WithLatests returns a fresh copy of n, with Metadata m merged in.
 func (n Node) WithLatests(m map[string]string) Node {
-	result := n.Copy()
 	ts := mtime.Now()
 	for k, v := range m {
-		result.Latest = result.Latest.Set(k, ts, v)
+		n.Latest = n.Latest.Set(k, ts, v)
 	}
-	return result
+	return n
 }
 
 // WithLatest produces a new Node with k mapped to v in the Latest metadata.
 func (n Node) WithLatest(k string, ts time.Time, v string) Node {
-	result := n.Copy()
-	result.Latest = result.Latest.Set(k, ts, v)
-	return result
+	n.Latest = n.Latest.Set(k, ts, v)
+	return n
 }
 
 // WithCounters returns a fresh copy of n, with Counters c merged in.
 func (n Node) WithCounters(c map[string]int) Node {
-	result := n.Copy()
-	result.Counters = result.Counters.Merge(Counters{}.fromIntermediate(c))
-	return result
+	n.Counters = n.Counters.Merge(Counters{}.fromIntermediate(c))
+	return n
 }
 
 // WithSet returns a fresh copy of n, with set merged in at key.
 func (n Node) WithSet(key string, set StringSet) Node {
-	result := n.Copy()
-	result.Sets = result.Sets.Add(key, set)
-	return result
+	n.Sets = n.Sets.Add(key, set)
+	return n
 }
 
 // WithSets returns a fresh copy of n, with sets merged in.
 func (n Node) WithSets(sets Sets) Node {
-	result := n.Copy()
-	result.Sets = result.Sets.Merge(sets)
-	return result
+	n.Sets = n.Sets.Merge(sets)
+	return n
 }
 
 // WithMetric returns a fresh copy of n, with metric merged in at key.
 func (n Node) WithMetric(key string, metric Metric) Node {
-	result := n.Copy()
-	result.Metrics[key] = n.Metrics[key].Merge(metric)
-	return result
+	n.Metrics = n.Metrics.Copy()
```

Comment:
Worth commenting that this works because n is passed by copy? (I don't think it's completely obvious, it took me a while to understand what was being done)